kernel/src/hil/symmetric_encryption.rs

```diff
@@ -3,40 +3,57 @@
 // Copyright Tock Contributors 2022.
 
 //! Interface for symmetric-cipher encryption
-//!
-//! see boards/imix/src/aes_test.rs for example usage
+//! (TODO) ADD EXAMPLE USAGE POINTER
 
-use crate::ErrorCode;
+use crate::{
+    utilities::leasable_buffer::{SubSlice, SubSliceMut},
+    ErrorCode,
+};
+
+pub enum Mode {
+    ECB,
+    CBC,
+    CTR,
+    CCM,
+    GCM,
+}
 
 /// Implement this trait and use `set_client()` in order to receive callbacks from an `AES128`
-/// instance.
-pub trait Client<'a> {
-    fn crypt_done(&'a self, source: Option<&'static mut [u8]>, dest: &'static mut [u8]);
+/// instance. This returns the provided source. The `dest` contains the result of the operation
+/// and in both cases the `SubSlic` provided to `crypt()` will be returned.
+pub trait Client {
+    fn crypt_done(
+        &self,
+        source: Option<SubSlice<'static, u8>>,
+        dest: Result<SubSliceMut<'static, u8>, (ErrorCode, SubSliceMut<'static, u8>)>,
+    );
 }
 
 /// The number of bytes used for AES block operations.  Keys and IVs must have this length,
 /// and encryption/decryption inputs must be have a multiple of this length.
 pub const AES128_BLOCK_SIZE: usize = 16;
 pub const AES128_KEY_SIZE: usize = 16;
 
-pub trait AES128<'a> {
+pub trait AES128Hw<'a> {
     /// Enable the AES hardware.
-    /// Must be called before any other methods
+    /// Must be called before any other methods.
     fn enable(&self);
 
-    /// Disable the AES hardware
+    /// Disable the AES hardware.
     fn disable(&self);
```

Comment:
Do we actually need `enable/disable` methods? From a quick (and likely imperfect) code search, I think the only thing that is using this currently is the `sam4l` to power on and off peripheral clocks, which isn't something that should be in the generic HIL, but rather a detail abstracted to the driver implementation.

While `enable`/`disable` isn't quite covered by the [avoid `initialize` in HILs](https://github.com/tock/tock/blob/master/doc/reference/trd3-hil-design.md#rule-11-initialize-methods-when-needed-should-be-in-a-separate-trait-and-invoked-in-an-instantiating-component) specifically, I think it's similar in spirit.